tket/src/Mapping/MappingFrontier.cpp

```diff
@@ -763,22 +788,39 @@ void MappingFrontier::add_bridge(
   // remove old cx vertex
   this->circuit_.remove_vertex(
       cx_v, Circuit::GraphRewiring::Yes, Circuit::VertexDeletion::Yes);
+  assert_valid_maps();
 }
 
 void MappingFrontier::add_ancilla(const UnitID& ancilla) {
-  Qubit qb(ancilla);
-  this->circuit_.add_qubit(qb);
-  this->linear_boundary->insert({qb, {this->circuit_.get_in(qb), 0}});
-
-  this->bimaps_->initial.insert({qb, qb});
-  this->bimaps_->final.insert({qb, qb});
+  // Given the ancilla node, we 1. add ancilla to the circuit.
+  // 2. map ancilla to placeholder_q, which will be a qubit from the routing
+  // ancilla register that is not present anywhere in the circuit and bimaps.
+  // Add {placeholder_q, ancilla} to the bimaps.
+  register_t ancilla_reg = this->circuit_.get_reg(q_routing_ancilla_reg());
+  unsigned available_ancilla_idx = 0;
+  auto it = ancilla_reg.rbegin();
+  if (it != ancilla_reg.rend()) {
+    available_ancilla_idx = it->first + 1;
+  }
+  for (auto const& it : this->bimaps_->initial.left) {
+    if (it.first.reg_name() == q_routing_ancilla_reg() &&
+        it.first.index()[0] >= available_ancilla_idx) {
+      available_ancilla_idx = it.first.index()[0] + 1;
+    }
+    if (it.second.reg_name() == q_routing_ancilla_reg() &&
+        it.second.index()[0] >= available_ancilla_idx) {
+      available_ancilla_idx = it.second.index()[0] + 1;
+    }
+  }
+  Qubit placeholder_q(q_routing_ancilla_reg(), available_ancilla_idx);
+  Qubit ancilla_q(ancilla);
+  this->circuit_.add_qubit(ancilla_q);
+  this->linear_boundary->insert(
+      {ancilla_q, {this->circuit_.get_in(ancilla_q), 0}});
+
+  this->bimaps_->initial.insert({placeholder_q, ancilla});
+  this->bimaps_->final.insert({placeholder_q, ancilla});
   this->ancilla_nodes_.insert(Node(ancilla));
-  UnitID uid_ancilla(ancilla);
-
-  unit_map_t update_map;
-  update_map.insert({uid_ancilla, uid_ancilla});
-
-  update_maps(this->bimaps_, update_map, update_map);
```

Comment:
I'm not sure if this actually does anything.

Reply:
I think if you're directly modifying this->bimaps_->initial/final yourself then we can probably avoid this? Also if tests are passing then seems fine to remove.